src/v/cluster/data_migrated_resources.cc

```diff
@@ -0,0 +1,171 @@
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "cluster/data_migrated_resources.h"
+
+#include "container/chunked_hash_map.h"
+#include "data_migration_types.h"
+
+#include <seastar/util/variant_utils.hh>
+
+#include <absl/container/flat_hash_set.h>
+
+namespace cluster {
+
+namespace {
+template<typename T>
+void remove_from_resources(
+  data_migration_id id,
+  const T& to_remove,
+  chunked_hash_map<T, data_migrated_resources::resource_metadata>& resources) {
+    auto it = resources.find(to_remove);
+    vassert(
+      it != resources.end(),
+      "migrated {} resource must exists in migrated resources");
+    vassert(
+      it->second.migration_id == id,
+      "removed migration must match migrated resource migration id");
+    resources.erase(it);
+}
+
+migrated_resource_state
+get_outbound_migration_resource_state(data_migration_state state) {
+    switch (state) {
+    case data_migration_state::planned:
+    case cluster::data_migration_state::preparing:
+    case cluster::data_migration_state::prepared:
+    case cluster::data_migration_state::canceling:
+    case cluster::data_migration_state::cancelled:
+        return migrated_resource_state::restricted;
+    case cluster::data_migration_state::executing:
+    case cluster::data_migration_state::executed:
+        return migrated_resource_state::blocked;
+    case cluster::data_migration_state::finished:
+        return migrated_resource_state::non_restricted;
+    }
+}
+
+} // namespace
+
+migrated_resource_state data_migrated_resources::get_topic_state(
+  const model::topic_namespace& tp_ns) const {
+    auto it = _topics.find(tp_ns);
+    if (it != _topics.end()) {
+        return it->second.state;
+    }
+    return migrated_resource_state::non_restricted;
+}
+
+migrated_resource_state
+data_migrated_resources::get_group_state(const consumer_group& cg) const {
+    auto it = _groups.find(cg);
+    if (it != _groups.end()) {
+        return it->second.state;
+    }
+    return migrated_resource_state::non_restricted;
+}
+
+void data_migrated_resources::apply_update(
+  const data_migration_metadata& migration_meta) {
+    ss::visit(
+      migration_meta.migration,
+      [this, id = migration_meta.id, state = migration_meta.state](
+        auto& migration) { apply_update(id, migration, state); });
+}
+
+void data_migrated_resources::apply_update(
+  data_migration_id id,
+  const outbound_data_migration& migration,
+  data_migration_state state) {
+    auto target_state = get_outbound_migration_resource_state(state);
+    for (const auto& t : migration.topics) {
+        auto [it, _] = _topics.try_emplace(
+          t,
+          resource_metadata{
+            .migration_id = id,
+            .state = target_state,
+          });
+        it->second.state = target_state;
```

Comment:
```
auto [it, inserted] = _topics.try_emplace(...)
if (!inserted) {
    it->second.state = target_state;
}
```

is the the semantics you are going for? if so, i think it is ok to leave off the conditional, but there should be a comment like:

```
// noop for new inserts, update when insertion doesn't occur
it->second.state = target_state;
```

because in these semantics require that 

```
            .state = target_state,
          });
        it->second.state = target_state;
```

The first and last line in the snippet are always in sync. Just checking `inserted` makes that much clearer.